### ISSUE TITLE
Sentinel Faction Target Hook Broken

### ISSUE THREAD

**Issue by WhyDoWeWonder** [drtshock/Factions#1157]
FactionsUUID version: 1.6.9.5-U02.2-b119

Server software: git-TacoSpigot-8014601d 

Server: No errors, from Factions, this is just a problem with Sentinel hooking for faction targets.

Details: Hi, after taking a look at this plugin I really loved it and I wanted to switch over from MassiveCraft, the problem is my factions server also has mercenaries, which are NPC's with the sentinel trait that you can hire, sentinel faction targets are completely broken  and whenever a sentinel references it's faction targets, it starts going haywire. I believe the problem is in this class here:
https://github.com/mcmonkey4eva/Sentinel/blob/master/src/main/java/org/mcmonkey/sentinel/integration/SentinelFactions.java
I don't know if your plugin has anything similar I can use, if not please add it.

**Comment by LaxWasHere** [comment]
It should be up to your NPC plugin to support FUUID, contact the author.

**Comment by WhyDoWeWonder** [comment]
I contacted him before I posted my thread here, he said to talk to you guys because it was your lack of similar hooks in your plugin that didn't allow him to do anything.

**Comment by drtshock** [comment]
If he tells me what I need to support, I can add support for it.

**Comment by mcmonkey4eva** [comment]
The required thing to support was linked in the opening post.
This file: https://github.com/mcmonkey4eva/Sentinel/blob/master/src/main/java/org/mcmonkey/sentinel/integration/SentinelFactions.java
Is the entirety of Factions linkage relevant.

It looks (from a glance at the github repo) like this project is heavily reorganized compared to Factions itself and essentially not equivalent to Factions (and thus probably shouldn't be sharing a package name or plugin name, to properly indicate the differentiation).

**Comment by LaxWasHere** [comment]
The required documentation/support already exist. <https://github.com/drtshock/Factions/wiki/Developer-Help>

**Comment by mcmonkey4eva** [comment]
To clarify if there's confusion here:

Sentinel does not link to or support FactionsUUID.
Sentinel supports Factions.

This user is encountering an error because FactionsUUID claims to be Factions, but is not compatible as an API.

The reasonable solutions are for FactionsUUID to add cross-support classes (Classes that can be called equivalently to Factions and will redirect to the appropriate relevant FactionsUUID code), or for FactionsUUID to alter its package and plugin name to not be equivalent to a separate plugin.

**Comment by LaxWasHere** [comment]
FactionsUUID **_IS_** Factions, just a different version. It's up to developer to support such version with the provided API above.

**Comment by mcmonkey4eva** [comment]
That is not reasonable.

You cannot have an entirely distinct API while maintaining the same package and internal names.

That makes it impossible to support one project, the other, or both, without resorting to reflection backed project verification. Which, in the case of supporting Factions, makes no sense to do at all (why should a Factions supporting plugin have to verify that "Factions" *is* Factions (and thus is compatible), and not some other plugin that took the same name and package, as FactionsUUID has done?)

**Comment by drtshock** [comment]
Well that's just not true https://github.com/drtshock/ObsidianDestroyer

**Comment by drtshock** [comment]
I also want to let you know that this is a FORK of Factions, not an entirely new Factions plugin. The project was initially written by MassiveCraft folks, abandoned, and then picked up by me. Changing the name space when forking is:
1. kinda mean to remove the name of the original group
2. breaks all compatibility

While I agree that it would certainly be easier to depend on FUUID separately if we had a different name space, it's certainly not very difficult to do so in the current way and seeing as your project isn't trivial, I would assume you'd have the capability to accomplish this, especially given the example OSS solution that you can copy paste.

**Comment by mcmonkey4eva** [comment]
`Well that's just not true https://github.com/drtshock/ObsidianDestroyer`
You just wrote the phrase "that's just not true" followed by a link confirming my claim as true. That project uses reflection backed project verification, exactly as I stated would be the requirement.

`2. breaks all compatibility` you have already done that. Changing the name space acknowledges the lack of compatibility, and reduces errors like the one that occurred here:
A plugin (Sentinel) linked to Factions, and a user (the issue poster) loaded Sentinel on a server running FactionsUUID. As Sentinel does not link to or support FactionsUUID, it has no compatibility or intent of compatibility with FactionsUUID, but encountered an error because FactionsUUID triggered the system to attempt to load Factions support, which does not function as FactionsUUID **is not** Factions.

**Comment by mdcfe** [comment]
@mcmonkey4eva FactionsUUID is a fork of Factions 1.6.9.5, and is naturally not going to have (or even claim to have) the same API as MassiveCraft Factions 2.x. You can retrieve the plugin version string and check whether it contains `1.6.9.5-U` to load a different hook (or just to ignore FUUID), as that version string always identifies FactionsUUID and never MassiveCraft Factions.

**Comment by WhyDoWeWonder** [comment]
Though I'm not very experienced here, why not just rename all of your packages and what's the difficulty of just moving to the MassiveCraft Factions 2.x. API? After all, it would increase your user base with MassiveCraft factions becoming compatible, as most factions add-ons and hooks are designed for MassiveCraft Factions. As for ones that hook into the current API, the vast majority of them have FactionsUUID, and it would help the devs of those plugins a lot if they could make one plugin for MassiveFactions and FactionsUUID easily.

**Comment by LaxWasHere** [comment]
![https://i.imgur.com/pKVCdpE.png](https://i.imgur.com/pKVCdpE.png)

**Comment by mdcfe** [comment]
@WhyDoWeWonder Not a Factions or FactionsUUID dev here, but in simplified terms the API on the server is provided by whichever plugin is installed. If FUUID is installed instead of Factions 2.x, the 2.x API won't exist on the server and vice versa.

Adding support for the 2.x API isn't feasible, as that would likely require:
* Breaking every single plugin that currently intentionally hooks into FactionsUUID
* Breaking every plugin that supported Factions 1.6.9.5 (and therefore FactionsUUID to a limited extent)
* Requiring all FactionsUUID servers to also install MassiveCore
* Pretending to have exactly the same features as 2.x to plugins that hook in
  * Not allowing other plugins to access any FactionsUUID features that don't exist in Factions 2.x
  * Breaking other plugins that access any Factions 2.x features that don't exist in FactionsUUID

All of these mean that developers that want to actively support FactionsUUID will have to put in more effort, as suddenly things are going on behind the scenes to pretend that they're one and the same.

Essentially, FactionsUUID would have to pretend to *be* a version of Factions 2.x to support the same API, which is backwards and would just create confusion, let alone the fact that they aren't the same.

**TL;DR**: you wouldn't hire a French interpreter that only speaks German.

---

There isn't any reason that a dev can't support FactionsUUID and Factions 2.x just by checking the version string and following a different code path. If a dev doesn't want to do that, it's their problem, not drtshock's.

**Comment by mdcfe** [comment]
The corresponding Sentinel ticket is here: mcmonkey4eva/Sentinel#241

I've forked Sentinel at [N3FS/Sentinel](https://github.com/N3FS/Sentinel) to support FactionsUUID, but this hasn't been tested yet - you can download it from [the ender.zone build server](https://ci.ender.zone/job/Sentinel%20for%20FactionsUUID/lastSuccessfulBuild).

**Comment by WhyDoWeWonder** [comment]
I get it now man, thanks, especially for the fork.
